### ISSUE TITLE
Plugin circular dependencies issue

### ISSUE THREAD

**Issue by FraI3mega** [PaperSpigot/Paper#8867]
### Expected behavior

In the previous version, the plugins worked correctly. No circular dependencies errors were shown.

### Observed/Actual behavior

UltraSetHome plugin conflicts with Essentials even though they aren't co-dependent. Grief prevention is conflicting with a nonexistent plugin.

### Steps/models to reproduce

Install the newest fabric version(git-Paper-406), add Grief Prevention or  UltraSetHome plugin and start the server,

### Plugin and Datapack List

CreeperConfetti, Essentials, *LoginSecurity, LuckPerms, PluginConstructorAPI, *SimpleTpa, SleepFixer, Vault, UltraSetHome, Grief Prevention

### Paper version

[22:16:46 INFO]: Checking version, please wait...
[22:16:46 INFO]: This server is running Paper version git-Paper-406 (MC: 1.19.3) (Implementing API version 1.19.3-R0.1-SNAPSHOT) (Git: 7baf427)
You are running the latest version
Previous version: git-Paper-405 (MC: 1.19.3)

**Comment by Owen1212055** [comment]
Please provide logs.

**Comment by Machine-Maker** [comment]
**Please post logs as Owen said**, but
> Grief prevention is conflicting with a nonexistent plugin.

is probably `TheUnderground`. GriefPrevention has to release a new update. It's been fixed in the source code since July 2022, a new version hasn't been released by the author.

**Comment by benklop** [comment]
Seeing the same issue with the Magic plugin:

```
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] Circular dependencies detected!
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] You have a plugin that is depending on a plugin which refers back to that plugin. Your server will shut down until these are resolved, or the strategy is changed.
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] Circular dependencies:
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] Skript depends on Magic depends on Skript...
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] Magic depends on MobArena depends on Magic...
[23:41:27] [Server thread/ERROR]: [ConfiguredOrderedProviderStorage] If you would like to still load these plugins, acknowledging that there may be unexpected plugin loading issues, run the server with -Dpaper.useLegacyPluginLoading=true

```

I don't actually have / want Skript or MobArena.

**Comment by electronicboy** [comment]
it appears that the dependency chain is still created for plugins which aren't installed but are referred to by other plugins; I'm not 100% sure if this is intended, but, this is in part something you should take to plugin devs

**Comment by Machine-Maker** [comment]
The Magic plugin has an invalid plugin.yml. They are both depending on and requiring loading before MobArena. A plugin can't load after **and** before a plugin at the same time. Report this to the Magic plugin.

**Comment by benklop** [comment]
It looks like there is a SoftDepend on Skript and MobArena, and a LoadBefore or these two as well. As it's a SoftDepend, does that really imply load after?

**Comment by mbax** [comment]
Yes, softdepend means "If the plugin is there, I need to load after it"

**Comment by Machine-Maker** [comment]
> As it's a SoftDepend, does that really imply load after?

Yes absolutely. Imagine a plugin want's to only do extra stuff if "Multiverse" is installed. But works totally fine without it. Chances are the plugin also wants to have access to all worlds loaded by Multiverse. That requires being loaded after. So a softdepend is an optional dependency, that if present, should be loaded before the plugin.

**Comment by electronicboy** [comment]
> wouldn't it be sensible to also say 'if the plugin is not there, the question of load order regarding it is unimportant'?

Not really, no; The thing is still invalid and shouldn't be possible, ideally, we just maybe wanna change how this is presented to show that it's a meta issue that should be fixed

**Comment by bigpresh** [comment]
>  ideally, we just maybe wanna change how this is presented to show that it's a meta issue that should be fixed

Agreed - I think a lot of confusion could have been avoided if the error message was something like:

_"Plugin $plugin_name's plugin.yml specifies that it should be loaded before (via loadbefore) and after (via softdepends) another plugin $other_plugin - this is an invalid configuration and needs to be fixed by the developers of $plugin"_

(a bit wordy, perhaps, but would convey all the required information).

The error about being a circular dependency is, IMO, a bit misleading and obscures the true problem.

**Comment by DiscipleGeek** [comment]
>Agreed - I think a lot of confusion could have been avoided if the error message was something like:
>
>"Plugin $plugin_name's plugin.yml specifies that it should be loaded before (via loadbefore) and after (via softdepends) another plugin $other_plugin - this is an invalid configuration and needs to be fixed by the developers of $plugin"
>
>(a bit wordy, perhaps, but would convey all the required information).
>
>The error about being a circular dependency is, IMO, a bit misleading and obscures the true problem.

Also might be a good idea to not completely kill the server because of it. Some of these plugins with dependency issues are very old and not being updated.

**Comment by Owen1212055** [comment]
@DiscipleGeek Read the error, please. You can of course use the legacy loading system flag. This avoids the issue, but I’d highly recommend just getting your plugins updated. (Or to notify plugin authors)

https://docs.papermc.io/paper/reference/paper-plugins#cyclic-dependency-detection

**Comment by Owen1212055** [comment]
Changes have been done to resolve the majority of plugin introduced circular loading. However, it's still very much a thing that we want to enforce to prevent plugin loading issues.

You can, of course, enable the flag to skip this... but note that this is very much a temporary solution for you to resolve your plugin loading loops.

https://docs.papermc.io/paper/reference/paper-plugins#cyclic-plugin-loading
